lib/compress/reduce-vars.js

```diff
@@ -528,93 +539,149 @@ function handle_defined_after_hoist(parent) {
                 continue;
             }
 
-            // defun is hoisted, so always safe
+            // found a reference to another function
             if (
                 def.assignments === 0
                 && def.orig.length === 1
                 && def.orig[0] instanceof AST_SymbolDefun
             ) {
+                if (!defun_dependents_map.has(def.id)) {
+                    defun_dependents_map.set(def.id, []);
+                }
+                defun_dependents_map.get(def.id).push(fname_def.id);
                 continue;
             }
 
-            if (found_self_ref_in_other_defuns) {
-                def.fixed = false;
-                continue;
-            }
-
-            // for the slower checks below this loop
-            potential_conflicts.push({ defun, def, fname_def });
+            enclosing_defs.push(def);
             symbols_of_interest.add(def.id);
-            symbols_of_interest.add(fname_def.id);
-            defuns_of_interest.add(defun);
+        }
+
+        if (enclosing_defs.length) {
+            defun_enclosing_defs_map.set(fname_def.id, enclosing_defs);
         }
     }
 
-    // linearize all symbols, and locate defs that are read after the defun
-    if (potential_conflicts.length) {
-        // All "symbols of interest", that is, defuns or defs, that we found.
-        // These are placed in order so we can check which is after which.
-        const found_symbols = [];
-        // Indices of `found_symbols` which are writes
-        const found_symbol_writes = new Set();
-        // Defun ranges are recorded because we don't care if a function uses the def internally
-        const defun_ranges = new Map();
-
-        let tw;
-        parent.walk((tw = new TreeWalker((node, descend) => {
-            if (node instanceof AST_Defun && defuns_of_interest.has(node)) {
+    // No defuns use outside constants
+    if (!defun_enclosing_defs_map.size) {
+        return;
+    }
+
+    // `defun` id to array of enclsoing `defun` id
+    // reverse of `defun_dependents_map`, used to speed up updating `defun_first_read_map`
+    const enclosing_defuns_map = new Map();
```

Comment:
Is it possible to generate enclosing_defuns_map directly without this whole section?